components/style/atomic_refcell.rs

```diff
@@ -2,121 +2,358 @@
+const MAX_FAILED_BORROWS: usize = HIGH_BIT + (HIGH_BIT >> 1);
+
+struct AtomicBorrowRef<'b> {
+    borrow: &'b AtomicUsize,
+}
+
+impl<'b> AtomicBorrowRef<'b> {
+    #[inline]
+    fn new(borrow: &'b AtomicUsize) -> Self {
+        let new = borrow.fetch_add(1, atomic::Ordering::AcqRel) + 1;
+
+        // If the new count has the high bit set, panic. The specifics of how
+        // we panic is interesting for soundness, but irrelevant for real programs.
+        if new & HIGH_BIT != 0 {
+            if new == HIGH_BIT {
+                // We overflowed into the reserved upper half of the refcount
+                // space. Before panicking, decrement the refcount to leave things
+                // in a consistent immutable-borrow state.
+                //
+                // This will never happen in a real program.
+                borrow.fetch_sub(1, atomic::Ordering::AcqRel);
+                panic!("too many immutable borrows");
+            } else if new >= MAX_FAILED_BORROWS {
+                // During the mutable borrow, an absurd number of threads have
+                // incremented the refcount and panicked. To avoid hypothetically
+                // wrapping the refcount, we abort the process once a certain
+                // threshold is reached.
+                //
+                // This will never happen in a real program.
+                //
+                // FIXME(bholley): We can't currently whole-process abort in
+                // stable Rust, but it's coming soon. We should fix this when
+                // https://github.com/rust-lang/rust/issues/37838 hits release.
+                panic!("Too many failed borrows");
+            } else {
+                // This is the normal case, and the only one which should happen
+                // in a real program.
+                panic!("already mutably borrowed");
+            }
+        }
+
+        AtomicBorrowRef { borrow: borrow }
     }
 }
 
-impl<'a, T: 'a + Debug> Debug for AtomicRefMut<'a, T> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "{:?}", self.0.deref())
+impl<'b> Drop for AtomicBorrowRef<'b> {
+    #[inline]
+    fn drop(&mut self) {
+        let old = self.borrow.fetch_sub(1, atomic::Ordering::AcqRel);
+        // This assertion is technically incorrect in the case where another
+        // thread hits the hypothetical overflow case, since we might observe
+        // the refcount before it fixes it up (and panics). But that never will
+        // never happen in a real program, and this is a debug_assert! anyway.
+        debug_assert!(old & HIGH_BIT == 0);
     }
 }
 
-impl<T> AtomicRefCell<T> {
-    pub fn new(value: T) -> Self {
-        AtomicRefCell(RwLock::new(value))
+struct AtomicBorrowRefMut<'b> {
+    borrow: &'b AtomicUsize,
+}
+
+impl<'b> Drop for AtomicBorrowRefMut<'b> {
+    #[inline]
+    fn drop(&mut self) {
+        let old = self.borrow.swap(0, atomic::Ordering::AcqRel);
```

Comment:
You should use `store` instead of `swap` since the former will only generate a single store instruction instead of an atomic `xchg` instruction.